### PR TITLE
#118_Update of the modality endpoint in the student's Resume.

### PR DESCRIPTION
**Pull request by coxen34**
**Description:** This PR addresses the task described in the 
https://github.com/IT-Academy-BCN/ita-profiles-backend/issues/118 issue, which requested the update of the endpoint that returns the modality of a specific student's resume. Previously, this endpoint returned fake data, but with the changes introduced, it now returns real data.
### The specific changes are as follows:
1. The 'Modality' column has been added to the resume migration, so that it appears after 'specialization'.
2. Updated the corresponding template to cast to an array in the database.
3. The controller name has been changed and the corresponding factory and path have been updated.
4. Updated the documentation in Swagger to reflect these changes.
5. Unit tests and Feature tests have been added for the ModalityController and the ModalityService.

This functionality has been separated into a service layer:
```php 
App\Service\Student\ModalityService;
```

api/v1/modality/{studentId}

### CODE REVIEW

File: app/Annotations/OpenApi/controllersAnnotations/modalityAnnotations/AnnotationsModality.php [hunk 1]
```diff
@@ -1,24 +1,52 @@
...
-     *             type="array",
-     *             @OA\Items(type="string")
-     *         )
-     *     )
-     * )
-     */
-
+    /**
+ * @OA\Get(
+ *     path="/modality/{studentId}",
```
Comment (jordimorillo) [inline app/Annotations/OpenApi/controllersAnnotations/modalityAnnotations/AnnotationsModality.php:11]: If you change the url of your endpoint make sure to notify the frontend people who are adding this endpoint.

File: app/Http/Controllers/api/ModalityController.php [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+use App\Http\Controllers\Controller;
+use App\Models\Resume;
+
+
+class ModalityController extends Controller
+{
+    public function __invoke($studentId)
+    {
+        try {
+            $resume = Resume::where('student_id', $studentId)->first();
```
Comment (jordimorillo) [inline app/Http/Controllers/api/ModalityController.php:14]: Extract this logic from the controller. A controler should be only using a service and managing the output of application

**Review by jordimorillo — CHANGES_REQUESTED** [review]

File: app/Http/Controllers/api/ModalityController.php (hunk 2 above)
Comment (coxen34) [reply app/Http/Controllers/api/ModalityController.php:14]: Ok

File: app/Annotations/OpenApi/controllersAnnotations/modalityAnnotations/AnnotationsModality.php (hunk 1 above)
Comment (coxen34) [reply app/Annotations/OpenApi/controllersAnnotations/modalityAnnotations/AnnotationsModality.php:11]: Should I have created an issue to refactor the names?

Comment (jordimorillo) [reply app/Annotations/OpenApi/controllersAnnotations/modalityAnnotations/AnnotationsModality.php:11]: Probably. This is outside this issue scope.

File: app/Service/Student/ModalityService.php [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+class ModalityService
+{
+    public function execute($studentId)
+    {
+
+        $modality = $this->getModalityByStudentId($studentId);
+
+        if (!$modality) {
+            throw new \Exception('No se encontró el currículum del usuario', 404);
```
Comment (jordimorillo) [inline app/Service/Student/ModalityService.php:17]: You should throw a controlled exception and catch it in the controller

File: tests/Feature/Service/ModalityServiceTest.php [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        $this->assertEquals($resume->modality, $result);
+    }
+
+    public function testExecuteWithInvalidStudentId()
+    {
+        $nonExistentStudentId = 9999;
+        
+        $modalityService = new ModalityService();
+
+        $this->expectException(Exception::class);
```
Comment (jordimorillo) [inline tests/Feature/Service/ModalityServiceTest.php:44]: Replace with the new ModalityNotFoundException

File: tests/Feature/Student/ModalityTest.php [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+        $resume = Resume::where('student_id', $studentId)->first();
+
+        $this->assertNotNull($resume);
+
+        $response = $this->getJson(route('modality', ['studentId' => $studentId]));
+
+        $response->assertStatus(200);
+
+        $response->assertJsonStructure([
```
Comment (FranEnLaNube) [inline tests/Feature/Student/ModalityTest.php:33]: This is not in the name of the test

File: tests/Fixtures/Resumes.php
Comment (FranEnLaNube) [inline tests/Fixtures/Resumes.php:1]: Good idea to use these fixtures!

**Review by FranEnLaNube — COMMENTED** [review]
Great Job guys! 
I'd love how many tests you have made. There are a couple of them that I couldn't understand, but at least all of them have passed on my computer... :sweet_smile 
- I've made some smalls comments, any of them are really important.
- I've also tested the swagger website and is working properly.
:clap

**Review by FranEnLaNube — APPROVED** [review]

File: app/Exceptions/ModalityNotFoundException.php [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+namespace App\Exceptions;
+
+use Exception;
+use Throwable;
+
+class ModalityNotFoundException extends Exception
+{
+    public function __construct($studentId, $code = 0, Throwable $previous = null)
+    {
+        $message = 'No se encontrado ningún currículum para el estudiante con id: ' . $studentId;
```
[the remark this answers is not in the document]
Comment (jordimorillo) [reply app/Exceptions/ModalityNotFoundException.php:12]: Catalan.

File: app/Service/Student/ModalityService.php [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+use App\Exceptions\ModalityNotFoundException;
+
+
+class ModalityService
+{
+    public function execute($studentId)
+    {
+        $modality = $this->getModalityByStudentId($studentId);
+        return $modality;
+    }
```
[the remark this answers is not in the document]
Comment (coxen34) [reply app/Service/Student/ModalityService.php:20]: changed

File: app/Annotations/OpenApi/controllersAnnotations/modalityAnnotations/AnnotationsModality.php [hunk 8]
```diff
@@ -1,24 +1,60 @@
-<?php 
+<?php
+
```
Comment (jordimorillo) [inline app/Annotations/OpenApi/controllersAnnotations/modalityAnnotations/AnnotationsModality.php:2]:
```php
declare(strict_types=1);
```

File: app/Service/Student/ModalityService.php [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+class ModalityService
+{
+    public function execute($studentId)
+    {
+        return $this->getModalityByStudentId($studentId);
+        
+    }
+
+    public function getModalityByStudentId($studentId)
```
Comment (jordimorillo) [inline app/Service/Student/ModalityService.php:22]: Para simplificar esta función de la clase, podrías usar consultas más claras y evitar el uso de excepciones para controlar el flujo del programa. Aquí te dejo una versión más simplificada:

```php
public function getModalityByStudentId($studentId)
{
    $student = Student::find($studentId);

    if (!$student) {
        throw new StudentNotFoundException($studentId);
    }

    $resume = $student->resume()->first();

    if (!$resume) {
        throw new ModalityNotFoundException($studentId);
    }

    $modality = $resume->modality;

    if (!$modality) {
        throw new ModalityNotFoundException($studentId);
    }

    return $modality;
}
```

File: database/migrations/2024_04_09_093348_add_modality_to_resumes.php [hunk 10]
```diff
@@ -0,0 +1,29 @@
+<?php
+
```
Comment (jordimorillo) [inline database/migrations/2024_04_09_093348_add_modality_to_resumes.php:2]:
```php
declare(strinc_types=1);
```

**Review by jordimorillo — APPROVED** [review]
Os paso unos ultimos detalles. Corregidlo y ya podréis publicar.

**Review by Ric-De — APPROVED** [review]
Great work people. 

Everything works perfectly. My only suggestion is to fill the last part of Swagger documentation related to annotations Schemas.

**Pull request merged**